components/equihash/tromp/equi_miner.c

```diff
@@ -0,0 +1,699 @@
+      return hash0[htl->prevhashunits-1].word == hash1[htl->prevhashunits-1].word;
+    }
+
+#if RESTBITS <= 6
+    typedef uchar xslot;
+#else
+    typedef u16 xslot;
+#endif
+  struct collisiondata {
+#ifdef XBITMAP
+#if NSLOTS > 64
+#error cant use XBITMAP with more than 64 slots
+#endif
+    u64 xhashmap[NRESTS];
+    u64 xmap;
+#else
+    xslot nxhashslots[NRESTS];
+    xslot xhashslots[NRESTS][XFULL];
+    xslot *xx;
+    u32 n0;
+    u32 n1;
+#endif
+    u32 s0;
+  };
+  typedef struct collisiondata collisiondata;
+
+    void collisiondata_clear(collisiondata *cd) {
+#ifdef XBITMAP
+      memset(cd->xhashmap, 0, NRESTS * sizeof(u64));
+#else
+      memset(cd->nxhashslots, 0, NRESTS * sizeof(xslot));
+#endif
+    }
+    bool addslot(collisiondata *cd, u32 s1, u32 xh) {
+#ifdef XBITMAP
+      xmap = xhashmap[xh];
+      xhashmap[xh] |= (u64)1 << s1;
+      s0 = -1;
+      return true;
+#else
+      cd->n1 = (u32)cd->nxhashslots[xh]++;
+      if (cd->n1 >= XFULL)
+        return false;
+      cd->xx = cd->xhashslots[xh];
+      cd->xx[cd->n1] = s1;
+      cd->n0 = 0;
+      return true;
+#endif
+    }
+    bool nextcollision(const collisiondata *cd) {
+#ifdef XBITMAP
+      return cd->xmap != 0;
+#else
+      return cd->n0 < cd->n1;
+#endif
+    }
+    u32 slot(collisiondata *cd) {
+#ifdef XBITMAP
+      const u32 ffs = __builtin_ffsll(cd->xmap);
+      s0 += ffs; cd->xmap >>= ffs;
+      return s0;
+#else
+      return (u32)cd->xx[cd->n0++];
+#endif
+    }
+
+  void equi_digit0(equi *eq, const u32 id) {
+    uchar hash[HASHOUT];
+    BLAKE2bState* state;
+    htlayout htl = htlayout_new(eq, 0);
+    const u32 hashbytes = hashsize(0);
+    for (u32 block = id; block < NBLOCKS; block += eq->nthreads) {
+      state = eq->blake2b_clone(eq->blake_ctx);
+      u32 leb = htole32(block);
+      eq->blake2b_update(state, (uchar *)&leb, sizeof(u32));
+      eq->blake2b_finalize(state, hash, HASHOUT);
+      eq->blake2b_free(state);
+      for (u32 i = 0; i<HASHESPERBLAKE; i++) {
+        const uchar *ph = hash + i * WN/8;
+#if BUCKBITS == 16 && RESTBITS == 4
+        const u32 bucketid = ((u32)ph[0] << 8) | ph[1];
+#elif BUCKBITS == 12 && RESTBITS == 8
+        const u32 bucketid = ((u32)ph[0] << 4) | ph[1] >> 4;
+#elif BUCKBITS == 11 && RESTBITS == 9
+        const u32 bucketid = ((u32)ph[0] << 3) | ph[1] >> 5;
+#elif BUCKBITS == 20 && RESTBITS == 4
+        const u32 bucketid = ((((u32)ph[0] << 8) | ph[1]) << 4) | ph[2] >> 4;
+#elif BUCKBITS == 12 && RESTBITS == 4
+        const u32 bucketid = ((u32)ph[0] << 4) | ph[1] >> 4;
+        const u32 xhash = ph[1] & 0xf;
+#else
+#error not implemented
+#endif
+        const u32 slot = getslot(eq, 0, bucketid);
+        if (slot >= NSLOTS) {
+          eq->bfull++;
+          continue;
+        }
+        slot0 *s = &eq->hta.trees0[0][bucketid][slot];
+        s->attr = tree_from_idx(block * HASHESPERBLAKE + i);
+        memcpy(s->hash->bytes+htl.nextbo, ph+WN/8-hashbytes, hashbytes);
```

Comment:
Currently the solver segfaults unless I comment out the second and third of these lines (and similarly comment out any accesses to `htlayout` data). I suspect that means my porting of the code from C++ to C is buggy.